storage/command_queue.go

```diff
@@ -87,17 +95,71 @@ func (cq *CommandQueue) GetWait(readOnly bool, wg *sync.WaitGroup, spans ...roac
 			end = start.Next()
 			start = end[:len(start)]
 		}
-		for _, c := range cq.cache.GetOverlaps(start, end) {
-			c := c.Value.(*cmd)
-			// Only add to the wait group if one of the commands isn't read-only.
-			if !readOnly || !c.readOnly {
+		overlaps := cq.cache.GetOverlaps(start, end)
+		if readOnly {
+			// If both commands are read-only, there are no dependencies between them,
+			// so these can be filtered out of the overlapping commands.
+			overlaps = filterReadWrite(overlaps)
+		}
+
+		// Easy cases that can avoid sorting and allocation.
+		switch len(overlaps) {
+		case 0:
+			continue
+		case 1:
+			c := overlaps[0].Value.(*cmd)
+			c.pending = append(c.pending, wg)
+			wg.Add(1)
+			continue
+		}
+
+		// Sort overlapping commands by timestamp and iterate from latest to earliest,
+		// adding the commands' ranges to the RangeGroup to determine gating keyspace
+		// command dependencies. Because all commands are given WaitGroup dependencies
+		// to the most recent commands that they are dependent on, and because of the
+		// causality provided by the Hybrid Logical Clock used to timestamp commands,
+		// this approach will construct a DAG-like dependency graph between WaitGroups
+		// with overlapping keys. This comes as an alternative to creating explicit
+		// WaitGroups dependencies to all gating commands for each new command, which
+		// could result in an exponential dependency explosion.
+		sort.Sort(overlapSorter(overlaps))
+		for i := len(overlaps) - 1; i >= 0; i-- {
+			c := overlaps[i]
+			if cq.rg.Add(c.Key.Range) {
+				c := c.Value.(*cmd)
 				c.pending = append(c.pending, wg)
 				wg.Add(1)
 			}
 		}
+		cq.rg.Clear()
 	}
 }
 
+// filterReadWrite filters out the read-only commands from the provided slice.
+func filterReadWrite(cmds []cache.Overlap) []cache.Overlap {
+	rwIdx := len(cmds)
+	for i := 0; i < rwIdx; {
+		c := cmds[i].Value.(*cmd)
+		if !c.readOnly {
+			i++
+		} else {
+			cmds[i], cmds[rwIdx-1] = cmds[rwIdx-1], cmds[i]
```

Comment:
ah, nevermind. I was picturing this having to do with spans passed into beginCmd, but it's not the case.